### ISSUE TITLE
Add extra Galactic coordinate frames in Astropy?

### ISSUE THREAD

**Issue by cdeil** [astropy/astropy#3344]
As discussed in various other issues already (e.g. https://github.com/astropy/coordinates-benchmark/issues/37, https://github.com/astropy/astropy/pull/3107), Galactic coordinates are not well-defined and different codes / catalogs use different definitions.

E.g. @adonath and I are currently producing a source catalog for [HESS](https://www.mpi-hd.mpg.de/hfm/HESS/) using the current Astropy Galactic frame and noticed that it differs by ~ 2 arcsec from the Galactic frame used in the [Fermi catalogs](http://fermi.gsfc.nasa.gov/ssc/data/access/lat/4yr_catalog/) ... see the check [here](http://nbviewer.ipython.org/github/gammapy/gammapy-extra/blob/master/checks/galactic-coordinate-check.ipynb).

* What definition for Galactic coordinates is most commonly used for source catalogs?
* Which definition do NED / Simbad / Vizier use when they use Galactic coordinates?

And here's my feature request for Astropy:
* Can we please add the one or two alternative Galactic frames that are in use in the astronomy community to `astropy.coordinates`?

I'm not sure if it would be best to do this with classes / functions that are part of the public API or if we want to keep this as internal classes / functions. But I think it would be useful to have at least to check / reproduce Galactic coordinate results from others.

@astrofrog @eteq @dsberry, all: what do you think?

**Comment by astrofrog** [comment]
@cdeil - is the 2" offset with the latest master or 0.4?

**Comment by astrofrog** [comment]
Currently, all the different Python codes currently agree to within << 1", and even astrolib agrees within 0.2" with other codes. I think we also agree with ds9. I've never seen a 2" offset so I actually think something may be wrong in the way the Galactic coordinates were computed in the Fermi catalogs? Is there a way to ask them what they used?

**Comment by cdeil** [comment]
The 2 arcsec offset is with the latest Astropy master.
I've written an email to the Fermi catalog contact ... at the moment I don't know what they use.

I still don't understand the details of the various Galactic systems, but if I'm not mistaken Astropy and most other codes do something slightly different from [Reid & Brunthaler 2004](http://adsabs.harvard.edu/abs/2004ApJ...616..872R) or what is proposed in [Liu et al. 2011](http://adsabs.harvard.edu/abs/2011A%26A...526A..16L).

IMO the main criterion whether it makes sense to add other Galactic frames to Astropy is whether they are used in practice, not if they are different by 2 arcsec or 0.2 arcsec.
Has anyone checked other source catalogs that publish both RaDec and Galactic positions or know what NED / SIMBAD / Vizier use (I presume they have code to compute Galactic coordinates for catalogs that don't include Galactic position columns themselves)?

**Comment by astrofrog** [comment]
If there are two clear competing definitions then I agree we can consider whether to add additional frames, but I'm going to guess that for every catalog you find you will find a slightly different convention used. However, I'm willing to be proved wrong! But if Fermi aare the only ones to use their convention I don't think it justifies the introduction of a new frame. +1 to checking what Vizier do.

**Comment by astrofrog** [comment]
This thread is relevant:

http://mail.scipy.org/pipermail/astropy/2014-November/003510.html

Also:

http://mail.scipy.org/pipermail/astropy/2014-November/003523.html

**Comment by cdeil** [comment]
So at the very least we have to document what is used in Astropy ... I think I mentioned this already somewhere but now there's an open issue so we don't forget. :-)

**Comment by eteq** [comment]
Yeah, it sounds like the one we're using (galactic is defined by poles in FK4 with no E-terms) is the more "official", but it's a bit of a mess in the field so definitely @cdeil is right we should at least document it so that others know without having to dig into the mailing list.

Please let us know what (if anything) you hear from Fermi, ideally within a week so we can add this to the docs in time for 1.0.

**Comment by cdeil** [comment]
I did hear back from Fermi, but forgot to comment here.
They were using a self-written software to do this coordinate conversion at low precision  ... Fermi has a systematic error of ~ 10 arcsec, i.e. for them it didn't matter.
Still, just for consistency, now they plan to switch to use Astropy.

So all that's left to be done here is to add a sentence or two to the Astropy docs mentioning what exactly is implemented. I still don't understand the details and am not really sure what exactly to write so that it's well-defined and clear at least to the coordinate experts.

**Comment by astrofrog** [comment]
By the way, I think we really should look into the ERFA ICRS <-> Galactic conversion now since in the long term this is going to be the definitive 'exact' conversion to use. If anyone has time to look into it, it would be great to figure out how different that is from our current comparison.

**Comment by dsberry** [comment]
Presumably the new ERFA Galactic system will need to be seen as an
additional system rather than as a replacement for the existing Galactic
system, since people will have existing data sets calibrated in the old
system. The old system will need to be supported basically for ever in
order to allow legacy data sets to be interpreted correctly.

David

**Comment by astrofrog** [comment]
@dsberry - I am going to assume that the new system was fit to minimize deviations from the old system in most cases (at least that's how I would have defined it). I'd be fine with keeping the old system, but I think I'd like the ERFA version to be e.g. ``Galactic`` and the legacy one could be e.g. ``Galactic1958``.

In fact, the old system was poorly defined enough that no one can trust it to better than around 1" anyway, and there are many different implementations. So it's not like the current version we have in astropy is some kind of high accuracy legacy version in any case...

**Comment by astrofrog** [comment]
Another possibility is to consider the convention as a parameter to the Galactic frame, so ``Galactic()`` would give the new erfa frame, while ``Galactic(convention='1958')`` would give the original one (well, one interpretation of the original one, which is what we have currently in astropy)

**Comment by astrofrog** [comment]
If the differences between the old and new system is of the order of milliarseconds, this is moot anyway since the old system was never that accurate in absolute terms due to the different interpretations.

**Comment by dsberry** [comment]
I'm thinking of a case where someone has an image calibrated accurately in ICRS (for example), which is then transform into Galactic coordinates using the old convention.  They will need the old convention to be retained separately to the new convention so the image can if necessary be converted back from (inacccurate) old galactic coords to (accurate) ICRS at some point in the future. If they used the new convention, the final ICRS  map would not be quite the same as the original.

**Comment by Juanlu001** [comment]
![Resurrecting](https://media.giphy.com/media/EF7SPrk8gRtjq/giphy.gif)

This conversation is very similar to what we faced when adding new ecliptic frames:

https://github.com/astropy/astropy/pull/8394

We ended up adding one different class per type of frame, which was a bit controversial because increased the overall number of frames quite a bit, and possibly introduced a bit of confusion. I like the `Galactic(convention='1958')` idea that @astrofrog proposed above, but would be a bit complicated to implement with the current architecture. I have some ideas about this but it will take me some time.

**Comment by adrn** [comment]
::insert zombie resurrection gif here::

I'm open to using the ERFA convention, or adding the option to use the ERFA convention, but I don't want to use the ERFA transformation machinery itself to do this transformation. Right now, the transformation is implemented as a set of rotations, which make velocity (proper motion) transformations fast and analytic. If we switch to using the ERFA transformation directly, we would have to turn it into a `FunctionTransform` and the velocity transform would then have to be done via numerical derivatives through the function. But, since [ERFA provides the ICRS-->Galactic matrix](https://github.com/astropy/astropy/blob/master/cextern/erfa/g2icrs.c), we could adopt that when `convention='erfa'`. For reference, here's a comparison of positions in astropy's current Galactic frame transformed to ICRS compared to a frame that uses the ERFA definition: https://gist.github.com/0954aec46b2f7f581831b286e550cff9

Because of the nature of the frame transform graph, I don't think it would be easy to add a `convention='...'` frame attribute to control this, because one convention maps fk5<-->galactic and the other icrs<-->galactic. I don't think we can easily make a frame attribute "disable" a transformation pathway, so this would probably have to be implemented as a separate frame, but @eteq can correct me if I'm wrong.
